File: lec9/a.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void printArray(const int arr[], int size) {
	for (int i = 0; i < size; i++) {
		printf("%d ", arr[i]);
	}
	printf("\n");
}

void shuffleArray(int arr[], int size) {
	for (int i = size - 1; i > 0; i--) {
		int j = rand() % (i + 1);
		int temp = arr[i];
		arr[i] = arr[j];
		arr[j] = temp;
	}
}

int compareAscending(int a, int b) {
	return a > b;
}

int compareDescending(int a, int b) {
	return a < b;
}

void sortArray(int arr[], int size, int (*compare)(int, int)) {
	for (int i = 0; i < size - 1; i++) {
		for (int j = 0; j < size - i - 1; j++) {
			if (compare(arr[j], arr[j + 1])) {
				int temp = arr[j];
				arr[j] = arr[j + 1];
				arr[j + 1] = temp;
			}
		}
	}
}

int main(void) {
	int arr[] = {1, 2, 3, 4, 5};
	int size = sizeof(arr) / sizeof(arr[0]);

	srand((unsigned int)time(NULL));

	printf("원본 배열: ");
	printArray(arr, size);

	shuffleArray(arr, size);
	printf("섞인 배열: ");
	printArray(arr, size);

	sortArray(arr, size, compareAscending);
	printf("오름차순 정렬 배열: ");
	printArray(arr, size);

	sortArray(arr, size, compareDescending);
	printf("내림차순 정렬 배열: ");
	printArray(arr, size);

	return 0;
}

File: lec9/a.py
import random
import time


def print_array(arr):
    print(*arr)


def shuffle_array(arr):
    for i in range(len(arr) - 1, 0, -1):
        j = random.randint(0, i)
        arr[i], arr[j] = arr[j], arr[i]


def compare_ascending(a, b):
    return a > b


def compare_descending(a, b):
    return a < b


def sort_array(arr, compare):
    size = len(arr)
    for i in range(size - 1):
        for j in range(size - i - 1):
            if compare(arr[j], arr[j + 1]):
                arr[j], arr[j + 1] = arr[j + 1], arr[j]


def main():
    arr = [1, 2, 3, 4, 5]

    random.seed(time.time())

    print("원본 배열:", end=" ")
    print_array(arr)

    shuffle_array(arr)
    print("섞인 배열:", end=" ")
    print_array(arr)

    sort_array(arr, compare_ascending)
    print("오름차순 정렬 배열:", end=" ")
    print_array(arr)

    sort_array(arr, compare_descending)
    print("내림차순 정렬 배열:", end=" ")
    print_array(arr)


if __name__ == "__main__":
    main()
